mcode/mcode-coverage-checker
# Refactoring coverage computations

# Description
Issue: STEAM-1068

This PR changes how coverage is calculated. Previously we would sum up all instances of a field being covered, which led to weird issues like when no resources of a certain profile were present, that would not contribute to the overall coverage in any way leading to falsely inflated percentages. This PR introduces a fixed list of fields to be checked for coverage and only considers a field covered if at least one resource has that field covered. This means that each fields coverage can only be 0 or 1, meaning the overall coverage is now a fraction with a fixed denominator of the total number of fields we check for. Before, resources with many instances would contribute more to the denominator of the overall coverage and totally missing profiles would contribute nothing. I have also kept the old behavior around and used it in the Rankings section because I think it makes sense there, I'll explain more below. Let me know if you have any questions, it's definitely kind of confusing.

Here are some examples of how the app should now work to help explain:
- The coverage of a profile viewed in the subcategory table now has a fixed range between 0 and the number of fields possible (e.g. Cancer Patient coverage only ranges from 0 to 8, before if there were 2 instances of this resource you could have gotten between 0 and 16)
- The coverage of an individual field in the subcategory table can only between 0 and 1, if any one instance of this profile has the field covered than it is considered covered, if no instance of the profile has it, or there are no instances of that profile at all, it is uncovered (so we can now safely switch to the checkmarks originally used in the mockup, but I haven't added that yet in this PR)
- The Rankings table still uses the old method of coverage counting. For example on `bundle1.json`, you could look at the Medication field of Cancer Related Medication Request in the subcategory table and see that it's covered, then go over to the Rankings section and see the additional information that all 4 instances of that field are covered. So now the rankings section provides additional info of how many instances of your covered fields are actually covered.

## Important Changes

_General changes_

`coverageFieldIds.js`
- We previously didn't have all fields explicitly declared anywhere, which led to weird omissions or 0/0 counts, so now this file will act as the source for a list of all fields we check for
- I went with an object where the keys are profile names and the values are arrays of field names

`statUtils.js`
- A lot of the functions here have been refactored to match the behavior described above
- Since I kept the old method around, there are basically two version of each function
  - The function names that use the term `Count` are the new method where coverage is binary
  - The function names that use the term `Sum` are the old method that multiplies with multiple instances of resources
  - I tried to write very descriptive comments over each function to make it clear which is which cuz it's definitely confusing
 
`Rankings.js`
- No actual behavior changes here, just changed the name of a function as described above

## Testing Recommendations
- Make sure the app now functions as described above
- Check the examples I gave above and see that they make sense
- Compare overall percentages from the previous version of the app with this version and make sure the new percentages better reflect coverage (e.g. `bundle2.json` now has a much lower percentage, which is more accurate)

# Checklists

**Submitter:**

@ACCT1 :
